<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Martin Petr">
<meta name="author" content="CC BY 4.0">

<title>Building intuition into popgen concepts and simulation-based inference</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="exercises_files/libs/clipboard/clipboard.min.js"></script>
<script src="exercises_files/libs/quarto-html/quarto.js"></script>
<script src="exercises_files/libs/quarto-html/popper.min.js"></script>
<script src="exercises_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="exercises_files/libs/quarto-html/anchor.min.js"></script>
<link href="exercises_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="exercises_files/libs/quarto-html/quarto-syntax-highlighting-e26003cea8cd680ca0c55a263523d882.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="exercises_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="exercises_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="exercises_files/libs/bootstrap/bootstrap-973236bd072d72a04ee9cd82dcc9cb29.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
</div>
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Building intuition into popgen concepts and simulation-based inference</h1>
<p class="subtitle lead"><a href="http://evomics.org/workshops/workshop-on-population-and-speciation-genomics/2025-workshop-on-population-and-speciation-genomics-cesky-krumlov/">Workshop on population and speciation genomics</a></p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Authors</div>
    <div class="quarto-title-meta-contents">
             <p>Martin Petr </p>
             <p><a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a> </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="reference-materials" class="level1">
<h1>Reference materials</h1>
<p>Unless you’re already a <em>slendr</em> expert, you will probably need some reference materials to do the exercises. Here are the options:</p>
<ol type="1">
<li>You can refer to the slides with the <em>slendr</em> crash course. The exercises will follow the progression of our walkthrough of the <em>slendr</em> functionality in the first part of this activity. You can find the material rendered as [normal slides]() or [continuous handouts]() (the latter being probably a bit more practical for reference.</li>
<li>Tutorials on the <a href="https://www.slendr.net"><em>slendr</em> website</a>.</li>
<li><a href="https://www.slendr.net/reference/index.html">Manual pages</a> of all available <em>slendr</em> functions. Note that you can get the help page of every <em>slendr</em> R <code>function</code> by typing <code>?function</code> in the R console. For instance, typing <code>?ts_tajima</code> gives you the help page of the <em>tskit</em>/<em>slendr</em> function implementing the tree-based computation of Tajima’s D.</li>
</ol>
</section>
<section id="installation-setup" class="level1">
<h1>Installation setup</h1>
<p>The easiest way to set up everything on your computer is to do the following:</p>
<ol type="1">
<li><p>Clone the repository with the activity materials (source code with slides and exercises materials, example scripts, and solutions). In a shell terminal, in your home directory (or anywhere else, really) you can run:</p>
<pre><code>$ git clone https://github.com/bodkan/cesky-krumlov-2025 ~/slendr_activity</code></pre></li>
<li><p>Install all the R package dependencies by going into the activity repository directory you just cloned and installing the necessary R packages. We’re using the <em>renv</em> package to manage all the dependencies, so you don’t have to track every package individually.</p></li>
</ol>
<ul>
<li><p>First go into the project directory:</p>
<pre><code>$ cd ~/slendr_activity</code></pre></li>
<li><p>Open the R terminal in that directory. You should get a note that the renv package is being automatically setup, like this:</p>
<pre><code>$ R

[... R startup information stuff ...]

# Bootstrapping renv 1.0.11 --------------------------------------------------
- Downloading renv ... OK
- Installing renv  ... OK

- Project '~/slendr_activity' loaded. [renv 1.0.11]
- One or more packages recorded in the lockfile are not installed.
- Use `renv::status()` for more details.</code></pre></li>
<li><p>Install the R package dependencies (still in the R console!):</p>
<pre><code>&gt; renv::restore(prompt = FALSE)</code></pre></li>
<li><p>Set up the Python environment used by the <em>slendr</em> R package for simulation and tree-sequence analysis (still in the R console!):</p>
<pre><code>&gt; slendr::setup_env(agree = TRUE)</code></pre>
<p>If everything worked, you should get an optimistic message saying:</p>
<pre><code>======================================================================
Python environment for slendr has been successfuly created, and the R
interface to msprime, tskit, and pyslim modules has been activated.

In future sessions, activate this environment by calling init_env().
=======================================================================</code></pre></li>
</ul>
<ol start="4" type="1">
<li>Open RStudio and navigate to your project directory via <code>File -&gt; Open Project...</code>.</li>
</ol>
<p><strong>If the <code>setup_env()</code> installation procedure fails, try the following:</strong></p>
<ol type="1">
<li>Delete the failed installation attempt:</li>
</ol>
<pre><code>slendr::clear_env(force = TRUE, all = TRUE)</code></pre>
<ol start="2" type="1">
<li>Try installing it again, this time using <code>pip</code> as a Python installation method (the default is <code>conda</code> which unfortunately fails fairly often):</li>
</ol>
<pre><code>slendr::setup_env(agree = TRUE, pip = TRUE)</code></pre>
<p>In every previous installments of this workshop, this is all that was needed to resolve problems.</p>
<p><strong>Installing SLiM</strong></p>
<p>It’s unclear whether we will manage to go through the entirety of the final exercise. However, to be able to do this, having SLiM at least version 4.2 (and it being available in your unix <code>$PATH!</code>) is required. If this isn’t possible for your, don’t worry. You’ll be able to do most of that exercise even without SLiM, and I will demonstrate the whole exercise (including the SLiM bit) for you.</p>
</section>
<section id="how-will-the-exercises-work" class="level1">
<h1>How will the exercises work?</h1>
<p>Your goal for each exercise will be to write a complete script. I suggest you name the script for each exercise as <code>exercise1.R</code>, <code>exercise2.R</code>, etc., to keep things tidy.</p>
<p>Unless you have a strong preference for another editor or IDE, I strongly suggest you use RStudio (either on your machine or in the cloud, depending on where you did the setup steps above).</p>
<p>Each exercise is compose of individual <em>parts</em>, which are designed to build one upon the other in the order they are specified.</p>
<p>All the exercises will involve “real coding”! If you’ve never really programmed entire scripts before, this could feel a little intimidating. Don’t worry. If you’re ever lost, just take a peek into the <a href="solutions/"><code>solutions/</code></a> directory for the respective solution script. Always try to work on a solution on your own, but never let this be a barrier to your progress. Feel free to copy-paste bits of my solutions into your own scripts.</p>
<p>If you find yourself <a href="https://scryfall.com/card/gtc/54/totally-lost"><em>totally lost</em></a>, don’t hesitate to read my solution scripts from the get go, trying to understand what’s going on in them, and executing them line by line in your RStudio session.</p>
</section>
<section id="exercise-1" class="level1 page-columns page-full">
<h1>Exercise 1</h1>
<p><strong>Reminder</strong>: you can use the <a href="">handout version of the slides</a> for reference.</p>
<section id="part-1-write-this-model-in-slendr" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="part-1-write-this-model-in-slendr">Part 1: write this model in <em>slendr</em></h2>
<p><img src="images/intro_model1.png" class="img-fluid" style="width:50.0%"></p>
<p><strong>Hint:</strong> Start script <code>exercise1.R</code> script in your RStudio session using this “template”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="sc">&lt;</span>... population definitions ...<span class="sc">&gt;</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="er">&lt;</span>... gene<span class="sc">-</span>flow definition ...<span class="sc">&gt;</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">populations =</span> <span class="fu">list</span>(...),</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_flow =</span> <span class="sc">&lt;</span>...<span class="sc">&gt;</span>,</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">30</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model) <span class="co"># verify visually</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> You can also specify time in units of “years ago”, just write them in a decreasing order (7Mya → 6Mya → …, as shown above).</p>
</div><div class="margin-aside">
<p><strong>Note:</strong> Plotting of models can be sometimes a little wonky. When plotting your model, experiment with arguments <code>log = TRUE</code>, <code>proportions = TRUE</code>, <code>gene_flow = TRUE</code>. Check <code>?plot_model</code> for more information on these.</p>
</div></div>
<section id="bonus" class="level3">
<h3 class="anchored" data-anchor-id="bonus">Bonus</h3>
<ul>
<li>Write your own model!</li>
</ul>
</section>
</section>
</section>
<section id="exercise-2" class="level1">
<h1>Exercise 2</h1>
<p><strong>Reminder</strong>: you can use the <a href="">handout version of the slides</a> for reference.</p>
<section id="sampling-adna-samples-through-time" class="level2">
<h2 class="anchored" data-anchor-id="sampling-adna-samples-through-time">Sampling aDNA samples through time</h2>
<p><br></p>
<p>Imagine we have <code>pop1</code>, <code>pop2</code>, … compiled in a <code>model</code>.</p>
<p><br></p>
<p>To record <em>ancient</em> individuals in the tree sequence, we can use <code>schedule_sampling()</code> like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">schedule_sampling</span>(</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  model,                <span class="co"># compiled slendr model object</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">100</span>, <span class="dv">500</span>),  <span class="co"># at these times (can be also a single number) ...</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop1, <span class="dv">42</span>),       <span class="co"># ... sample 42 individuals from pop1</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop2, <span class="dv">10</span>),       <span class="co"># ... sample 10 individuals from pop2</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop3, <span class="dv">1</span>)         <span class="co"># ... sample 1 individual from pop 3</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sampling-schedule-format" class="level2">
<h2 class="anchored" data-anchor-id="sampling-schedule-format">Sampling schedule format</h2>
<p>The output of <code>schedule_sampling()</code> is a plain data frame:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">40000</span>, <span class="dv">30000</span>, <span class="dv">20000</span>, <span class="dv">10000</span>), <span class="fu">list</span>(eur, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>. . .</p>
<p>We can bind multiple sampling schedules together, giving us finer control about sampling:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>eur_samples <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">40000</span>, <span class="dv">30000</span>, <span class="dv">20000</span>, <span class="dv">10000</span>, <span class="dv">0</span>), <span class="fu">list</span>(eur, <span class="dv">1</span>))</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>afr_samples <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="dv">0</span>, <span class="fu">list</span>(afr, <span class="dv">1</span>))</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">rbind</span>(eur_samples, afr_samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="how-to-use-a-sampling-schedule" class="level2">
<h2 class="anchored" data-anchor-id="how-to-use-a-sampling-schedule">How to use a sampling schedule?</h2>
<p>To sample individuals based on a given schedule, we use the <code>samples =</code> argument of the <code>msprime()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msprime</span>(model, <span class="at">samples =</span> samples, <span class="at">sequence_length =</span> <span class="fl">1e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>. . .</p>
<p>We can verify that only specific individuals are recorded:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="exercise-4" class="level1 page-columns page-full">
<h1>Exercise #4</h1>
<p><br><br><br><br></p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p>Use <a href="https://bodkan.quarto.pub/ku-popgen2023-onepage/">these</a> one-page handouts for reference.</p>
</div></div><section id="exercise-4a-ancient-samples" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4a-ancient-samples">Exercise #4a: ancient samples</h2>
<p>Let’s return to your introgression model:</p>
</section>
<section id="exercise-4a-ancient-samples-1" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4a-ancient-samples-1">Exercise #4a: ancient samples</h2>
<p>Simulate data from your model using this sampling:</p>
<ul>
<li>one present-day CHIMP and AFR individual</li>
<li>20 present-day EUR individuals</li>
<li>1 NEA at 70 ky, 1 NEA at 40 ky</li>
<li>1 EUR every 1000 years between 50-5 kya</li>
</ul>
<p>Reminder: you can do this by:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="co"># rbind(...) together individual schedule_sampling() data frames</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msprime</span>(model, <span class="at">samples =</span> samples, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="exercise-4b-f_4-ratio-statistic" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4b-f_4-ratio-statistic">Exercise #4b: <span class="math inline">\(f_4\)</span>-ratio statistic</h2>
<p>Use <span class="math inline">\(f_4\)</span>-ratio statistic to replicate the <a href="https://www.pnas.org/doi/10.1073/pnas.1814338116#fig01">following figure</a>:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/neand_decline.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:60.0%"></p>
</figure>
</div>
</section>
<section id="exercise-3-more-statistics-a" class="level2">
<h2 class="anchored" data-anchor-id="exercise-3-more-statistics-a">Exercise #3: more statistics! (a)</h2>
<p>Use <code>msprime()</code> to simulate a 50Mb tree sequence <code>ts</code> from your introgression model in <code>model1.R</code> (if that takes more than two minutes, try just 10Mb).</p>
<p>(Remember to add mutations with <code>ts_mutate()</code>.)</p>
</section>
<section id="exercise-3-more-statistics-b" class="level2">
<h2 class="anchored" data-anchor-id="exercise-3-more-statistics-b">Exercise #3: more statistics! (b)</h2>
<p><strong>In <code>model1.R</code> compute (some of) these on your <code>ts</code> object:</strong></p>
<ul>
<li><p>nucleotide <a href="https://www.slendr.net/reference/ts_diversity.html#ref-examples"><code>ts_diversity()</code></a> in each population <br></p></li>
<li><p><a href="https://www.slendr.net/reference/ts_divergence.html#ref-examples"><code>ts_divergence()</code></a> between populations<br></p></li>
<li><p>outgroup <a href="https://www.slendr.net/reference/ts_f4ratio.html#ref-examples"><code>ts_f3(A; B, C)</code></a> using CHIMP as the outgroup (A!) for different pairs of “B” and “C” populations<br></p>
<ul>
<li>using Ben’s explanation on Wednesday, try to compute this <span class="math inline">\(f_3\)</span> using combination of <span class="math inline">\(f_2\)</span> statistics (<code>ts_f2(A, B)</code>)</li>
</ul></li>
</ul>
<p><strong>You can find help by typing <code>?ts_diversity</code> etc. into R!</strong></p>
</section>
<section id="exercise-3-more-statistics-c" class="level2">
<h2 class="anchored" data-anchor-id="exercise-3-more-statistics-c">Exercise #3: more statistics! (c)</h2>
<ul>
<li><p><strong>Compute</strong> <span class="math inline">\(f_4\)</span> test of Neanderthal introgression in EUR:</p>
<ul>
<li><p>Hint: check the values of these two statistics (<code>ts_f4()</code>):</p>
<ul>
<li><span class="math inline">\(f_4\)</span>(&lt;afr&gt;, &lt;eur&gt;; &lt;neand&gt;, &lt;chimp&gt;)</li>
<li><span class="math inline">\(f_4\)</span>(&lt;afr1&gt;, &lt;afr2&gt;; &lt;neand&gt;, &lt;chimp&gt;)]</li>
</ul></li>
<li><p>Is one “much more negative” than the other as expected assuming introgression?</p></li>
</ul></li>
<li><p>You’ve learned about symmetries in <span class="math inline">\(f_4\)</span> depending on the arrangement of the “quartet”. <strong>How many unique</strong> <span class="math inline">\(f_4\)</span> values involving a single quartet can you find and why? (When computing <code>ts_f4()</code> here, set <code>mode = "branch"</code>).</p></li>
</ul>
</section>
</section>
<section id="exercise-3" class="level1 page-columns page-full">
<h1>Exercise 3</h1>
<p><strong>Reminder</strong>: you can use the <a href="">handout version of the slides</a> for reference.</p>
<p><br><br><br><br></p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p>Use <a href="https://bodkan.quarto.pub/ku-popgen2023-onepage/">these</a> one-page handouts for reference.</p>
</div></div><section id="part-a-one-population-afs-simulator" class="level2">
<h2 class="anchored" data-anchor-id="part-a-one-population-afs-simulator">Part <em>a</em>: One-population AFS simulator</h2>
<p>In a new script <code>model2.R</code> write a function called <code>simulate_afs()</code>, which will take <code>Ne</code> as its only parameter.</p>
<p>It should create a one-population <em>forward-time</em> model (<code>simulation_length</code> 100000, <code>generation_time</code> 1), simulate 10Mb tree sequence (recombination and mutation rates of 1e-8), compute AFS for 10 samples and return it.</p>
<p>. . .</p>
<hr>
<p><strong>Use this function to compute AFS vectors for various <code>Ne</code> values. Plot those AFS and observe how (and why?) do they differ based on <code>Ne</code> you simulated.</strong></p>
</section>
<section id="part-a-hint" class="level2">
<h2 class="anchored" data-anchor-id="part-a-hint">Part <em>a</em>: Hint</h2>
<p>You can start building <code>model2.R</code> from this “template”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr); <span class="fu">init_env</span>()</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>simulate_afs <span class="ot">&lt;-</span> <span class="cf">function</span>(Ne) {</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  ... your one<span class="sc">-</span>population model code<span class="sc">:</span> <span class="fu">population</span>(), <span class="fu">compile_model</span>(), <span class="fu">msprime</span>() ...</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>  result <span class="ot">&lt;-</span> ... compute AFS using <span class="fu">ts_afs</span>() on <span class="dv">10</span> samples, save it to <span class="st">`</span><span class="at">result</span><span class="st">`</span> ...</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(result)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>afs_1 <span class="ot">&lt;-</span> <span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">1000</span>)</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(afs_1, <span class="at">type =</span><span class="st">"o"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p><small> <strong>Note:</strong> We drop the first element (<code>afs[-1]</code>) technical reasons related to <em>tskit</em>. You don’t have to worry about that here, but you can read <a href="https://tskit.dev/tutorials/analysing_tree_sequences.html#sec-tutorial-afs-zeroth-entry">this</a> for more detail. </small></p>
</div>
<div class="fragment">
<p>When used in R, your function should work like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">1000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="part-b-estimating-ne-using-afs" class="level2">
<h2 class="anchored" data-anchor-id="part-b-estimating-ne-using-afs">Part <em>b</em>: Estimating Ne using AFS</h2>
<p>Imagine you sequenced 10 samples from a population and computed this AFS vector in R (# singletons, doubletons, etc.):</p>
<!-- dput(as.vector(observed_afs)) -->
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>afs_observed <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">2520</span>, <span class="dv">1449</span>, <span class="dv">855</span>, <span class="dv">622</span>, <span class="dv">530</span>, <span class="dv">446</span>, <span class="dv">365</span>, <span class="dv">334</span>, <span class="dv">349</span>, <span class="dv">244</span>,</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>                  <span class="dv">264</span>, <span class="dv">218</span>,  <span class="dv">133</span>, <span class="dv">173</span>, <span class="dv">159</span>, <span class="dv">142</span>, <span class="dv">167</span>, <span class="dv">129</span>, <span class="dv">125</span>, <span class="dv">143</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>. . .</p>
<p>You know that the population had a constant <span class="math inline">\(N_e\)</span> somewhere between 1000 and 30000 for the past 100,000 generations, and had mutation and recombination rates of 1e-8 (i.e., parameters already implemented by the <code>simulate_afs()</code> function).</p>
<p>. . .</p>
<hr>
<p><strong>Guess the true</strong> <span class="math inline">\(N_e\)</span> given the observed AFS by running single-population simulations for a range of <span class="math inline">\(N_e\)</span> values and comparing each run to <code>afs_observed</code>.</p>
</section>
<section id="part-b-hints" class="level2">
<h2 class="anchored" data-anchor-id="part-b-hints">Part <em>b</em>: Hints</h2>
<p>Using your custom <code>simulate_afs()</code> function, find the value of <code>Ne</code> that will give the closest AFS to the observed AFS:</p>
<ul>
<li><p><em>option #1</em> [<em>easy</em>]: Plot AFS vectors for various <span class="math inline">\(N_e\)</span> values, then eyeball which looks closest to the observed AFS</p></li>
<li><p><em>option #2</em> [<em>hard</em>]: Simulate AFS vectors in steps of possible <code>Ne</code> (maybe <code>lapply()</code>?), find the <a href="https://en.wikipedia.org/wiki/Mean_squared_error">closest</a> AFS</p></li>
</ul>
</section>
</section>
<section id="exercise-2-solution" class="level1 page-columns page-full">
<h1>Exercise #2: solution</h1>
<p><br><br><br><br></p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p>See <a href="https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex2_simple.R"><code>ex2_simple.R</code></a> for a simple “eyeballing” solution.<br></p>
<p>See <a href="https://github.com/bodkan/ku-popgen2023/blob/main/solutions/ex2_grid.R"><code>ex2_grid.R</code></a> for a more elaborate grid-search solution.</p>
</div></div></section>
<section id="exercise-4-1" class="level1">
<h1>Exercise 4</h1>
<p><strong>Reminder</strong>: you can use the <a href="">handout version of the slides</a> for reference.</p>
</section>
<section id="exercise-5" class="level1">
<h1>Exercise 5</h1>
<p><strong>Reminder</strong>: you can use the <a href="">handout version of the slides</a> for reference.</p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>