<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Martin Petr">
<meta name="author" content="CC BY 4.0">

<title>Building intuition into popgen concepts and simulation-based inference</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="slides_files/libs/clipboard/clipboard.min.js"></script>
<script src="slides_files/libs/quarto-html/quarto.js"></script>
<script src="slides_files/libs/quarto-html/popper.min.js"></script>
<script src="slides_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="slides_files/libs/quarto-html/anchor.min.js"></script>
<link href="slides_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="slides_files/libs/quarto-html/quarto-syntax-highlighting-e26003cea8cd680ca0c55a263523d882.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="slides_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="slides_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="slides_files/libs/bootstrap/bootstrap-973236bd072d72a04ee9cd82dcc9cb29.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
</div>
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Building intuition into popgen concepts and simulation-based inference</h1>
<p class="subtitle lead"><a href="http://evomics.org/workshops/workshop-on-population-and-speciation-genomics/2025-workshop-on-population-and-speciation-genomics-cesky-krumlov/">Workshop on population and speciation genomics</a></p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Authors</div>
    <div class="quarto-title-meta-contents">
             <p>Martin Petr </p>
             <p><a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a> </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="first-things-first" class="level1">
<h1>First things first</h1>
<p>These slides and other resources are (and always will be) at:</p>
<p><a href="https://github.com/bodkan/cesky-krumlov-2025">github.com/bodkan/cesky-krumlov-2025</a></p>
<p><br><br></p>
<p><strong>Open this link now so you have everything at hand later.</strong></p>
</section>
<section id="section" class="level1">
<h1></h1>
<div class="columns">
<div class="column" style="width:70%;">
<blockquote class="blockquote">
<p>Many problems in population genetics cannot be solved by a mathematician, no matter how gifted. [It] is already clear that computer methods are very powerful. This is good. It [‚Ä¶] <strong>permits people with limited mathematical knowledge to work on important problems</strong> [‚Ä¶].</p>
</blockquote>
</div><div class="column" style="width:30%;">
<p><img src="images/crow.jpeg" class="img-fluid"></p>
<p><a href="https://en.wikipedia.org/wiki/James_F._Crow">James F. Crow</a> ‚Äì <a href="http://www.gnxp.com/blog/2006/06/10-questions-for-jim-crow.php">interview</a></p>
</div>
</div>
</section>
<section id="why-use-simulations" class="level1 page-columns page-full">
<h1>Why use simulations?</h1>
<ol type="1">
<li>Making sense of inferred statistics</li>
<li>Fitting model parameters (i.e.&nbsp;<a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">ABC</a>)</li>
<li>Ground truth for method development</li>
</ol>
<section id="making-sense-of-inferred-statistics" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="making-sense-of-inferred-statistics">Making sense of inferred statistics</h2>
<center>
<img src="images/fstats_sims.png" class="img-fluid">
</center>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p>Image from <a href="https://academic.oup.com/genetics/article/202/4/1485/5930214">Peter (2016)</a></p>
</div></div></section>
<section id="making-sense-of-inferred-statistics-1" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="making-sense-of-inferred-statistics-1">Making sense of inferred statistics</h2>
<center>
<img src="images/admixture.png" class="img-fluid">
</center>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p>Image from <a href="https://www.nature.com/articles/s41467-018-05257-7">Lawson <em>et al.</em> (2018)</a></p>
</div></div></section>
<section id="fitting-model-parameters-i.e.-abc" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="fitting-model-parameters-i.e.-abc">Fitting model parameters (i.e.&nbsp;<a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">ABC</a>)</h2>
<center>
<img src="images/abc_scheme.png" class="img-fluid" style="width:50.0%">
</center>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p>Image from <a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">Wikipedia on ABC</a></p>
</div></div></section>
<section id="ground-truth-for-method-development" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="ground-truth-for-method-development">Ground truth for method development</h2>
<center>
<img src="images/mcmc.png" class="img-fluid">
</center>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p>Image from <a href="https://www.nature.com/articles/ng.3015">Schiffels and Durbin (2014)</a></p>
</div></div></section>
</section>
<section id="simulation-software" class="level1">
<h1>Simulation software</h1>
<p>The most famous and widely used are <a href="https://messerlab.org/slim/">SLiM</a> and <a href="https://tskit.dev/msprime/docs/stable/intro.html"><em>msprime</em></a>.</p>
<p>They are <u>very</u> powerful but both require:</p>
<div class="fragment">
<ul>
<li>quite a bit of programming knowledge,</li>
<li>a lot of code for non-trivial simulations (üêõü™≤üêú).</li>
</ul>
</div>
<div class="fragment">
<center>
<h3 class="anchored">
<strong>The exercises will focus on the <a href="http://www.slendr.net"><em>slendr</em></a></strong><br>popgen simulation toolkit for R.
</h3>
</center>
</div>
<div class="fragment">
<br>
<center>
But let‚Äôs look at SLiM and <em>msprime</em> at least a little bit.
</center>
</div>
<section id="section-1" class="level2">
<h2 class="anchored" data-anchor-id="section-1"></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2 class="anchored">
What is SLiM?
</h2>
<ul>
<li><strong>A forward-time simulator</strong></li>
</ul>
<div class="fragment">
<ul>
<li>It‚Äôs fully programmable!</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Massive library of functions for:
<ul>
<li>demographic events</li>
<li>various mating systems</li>
<li>selection, quantitative traits, ‚Ä¶</li>
</ul></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>&gt; 700 pages long <a href="https://github.com/MesserLab/SLiM/releases/download/v3.7.1/SLiM_Manual.pdf">manual</a>!</li>
</ul>
</div>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/sim_sketches.001.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption>Modified from <a href="http://alexeidrummond.org/bayesian_phylo_lectures/lecture10/">Alexei Drummond</a></figcaption>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="simple-neutral-simulation-in-slim" class="level2">
<h2 class="anchored" data-anchor-id="simple-neutral-simulation-in-slim">Simple neutral simulation in SLiM</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>initialize() {
    // create a neutral mutation type
    initializeMutationType("m1", 0.5, "f", 0.0);
    // initialize 1Mb segment
    initializeGenomicElementType("g1", m1, 1.0);
    initializeGenomicElement(g1, 0, 999999);
    // set mutation rate and recombination rate of the segment
    initializeMutationRate(1e-8);
    initializeRecombinationRate(1e-8);
}

// create an ancestral population p1 of 10000 diploid individuals
1 early() { sim.addSubpop("p1", 10000); }

// in generation 1000, create two daughter populations p2 and p3
1000 early() {
    sim.addSubpopSplit("p2", 5000, p1);
    sim.addSubpopSplit("p3", 1000, p1);
}

// in generation 10000, stop the simulation and save 100 individuals
// from p2 and p3 to a VCF file
10000 late() {
    p2_subset = sample(p2.individuals, 100);
    p3_subset = sample(p3.individuals, 100);
    c(p2_subset, p3_subset).genomes.outputVCF("/tmp/slim_output.vcf.gz");
    sim.simulationFinished();
    catn("DONE!");
}</code></pre>
</div>
</div>
</section>
<section id="section-2" class="level2">
<h2 class="anchored" data-anchor-id="section-2"></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2 class="anchored">
What is <em>msprime</em>?
</h2>
<ul>
<li>A Python module for writing <strong>coalescent simulations</strong></li>
</ul>
<div class="fragment">
<ul>
<li>Extremely fast (genome-scale, population-scale data)</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>You must know Python fairly well to build complex models</li>
</ul>
</div>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/sim_sketches.002.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption>Modified from <a href="http://alexeidrummond.org/bayesian_phylo_lectures/lecture10/">Alexei Drummond</a></figcaption>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="simple-simulation-using-msprime" class="level2">
<h2 class="anchored" data-anchor-id="simple-simulation-using-msprime">Simple simulation using <em>msprime</em></h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>import msprime

demography = msprime.Demography()
demography.add_population(name="A", initial_size=10_000)
demography.add_population(name="B", initial_size=5_000)
demography.add_population(name="C", initial_size=1_000)
demography.add_population_split(time=1000, derived=["A", "B"], ancestral="C")

ts = msprime.sim_ancestry(
  sequence_length=10e6,
  recombination_rate=1e-8,
  samples={"A": 100, "B": 100},
  demography=demography
)</code></pre>
</div>
</div>
</section>
</section>
<section id="section-3" class="level1">
<h1></h1>
<center>
<h2 class="anchored" data-anchor-id="section-3">
<a href="https://www.slendr.net">www.slendr.net</a>
</h2>
<p><img src="images/slendr_logo.png" class="img-fluid" style="width:30.0%"></p>
</center>
<div class="fragment">
<center>
<h2 class="anchored">
Why a new package?
</h2>
</center>
</div>
<section id="spatial-simulations" class="level2">
<h2 class="anchored" data-anchor-id="spatial-simulations">Spatial simulations!</h2>
<center>
<img src="images/animation.gif" class="img-fluid" style="width:70.0%">
</center>
</section>
<section id="section-4" class="level2">
<h2 class="anchored" data-anchor-id="section-4"></h2>
<h2 class="anchored">
Why a new package?
</h2>
<ul>
<li><p>Most researchers are not expert programmers</p></li>
<li><p>All but the most trivial simulations require lots of code</p></li>
</ul>
<div class="fragment">
<ul>
<li><p>90% <sup><font color="blue">[citation needed]</font></sup> of simulations are basically the same!</p>
<ul>
<li><p>create populations (splits and <span class="math inline">\(N_e\)</span> changes)</p></li>
<li><p>specify if/how they should mix (rates and times)</p></li>
</ul></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Lot of code duplication across projects</li>
</ul>
</div>
<div class="fragment">
<center>
<strong><em>slendr</em></strong> <strong>makes ‚Äústandard‚Äù demographic simulations trivial</strong><br>and unlocks new kinds of spatial simulations
</center>
</div>
</section>
</section>
<section id="lets-get-started" class="level1 page-columns page-full">
<h1>Let‚Äôs get started</h1>
<section id="everything-we-do-will-be-in-r" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="everything-we-do-will-be-in-r">Everything we do will be in R</h2>
<p><br><br></p>
<p>You can write your <em>slendr</em> code however you like, but I recommed using RStudio, for convenience.</p>
<p><br></p>
<p>Always start your R scripts with this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>The interface to all required Python modules has been activated.</code></pre>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p>(If you get a note about missing SLiM, safely ignore it for now.)</p>
</div></div></section>
<section id="typical-steps-of-a-slendr-r-workflow" class="level2">
<h2 class="anchored" data-anchor-id="typical-steps-of-a-slendr-r-workflow">Typical steps of a <em>slendr</em> R workflow</h2>
<p><br></p>
<ol type="1">
<li>creating populations</li>
<li>scheduling population splits</li>
<li>programming <span class="math inline">\(N_e\)</span> size changes</li>
<li>encoding gene-flow events</li>
<li>simulation sequence of a given size</li>
<li>computing statistics from simulated outputs</li>
</ol>
</section>
<section id="creating-a-population" class="level2">
<h2 class="anchored" data-anchor-id="creating-a-population">Creating a <code>population()</code></h2>
<p>Each needs a name, size and time of appearance:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>. . .</p>
<p><br></p>
<p>This creates a normal R object. Typing it out gives a summary:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>pop1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop1 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 1: created as an ancestral population (N = 1000)</code></pre>
</div>
</div>
</section>
<section id="programming-population-splits" class="level2">
<h2 class="anchored" data-anchor-id="programming-population-splits">Programming population splits</h2>
<p>Splits are indicated by the <code>parent = &lt;pop&gt;</code> argument:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">50</span>, <span class="at">parent =</span> pop1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>. . .</p>
<p><br></p>
<p>The split is reported in the ‚Äúhistorical summary‚Äù:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>pop2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop2 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 50: split from pop1 (N = 100)</code></pre>
</div>
</div>
</section>
<section id="scheduling-resize-events-resize" class="level2">
<h2 class="anchored" data-anchor-id="scheduling-resize-events-resize">Scheduling resize events ‚Äì <code>resize()</code></h2>
<p>Step size decrease:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb11-2"><a href="#cb11-2"></a>pop1_step <span class="ot">&lt;-</span> <span class="fu">resize</span>(pop1, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">how =</span> <span class="st">"step"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<p>Exponential increase:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>pop2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">50</span>, <span class="at">parent =</span> pop1)</span>
<span id="cb12-2"><a href="#cb12-2"></a>pop2_exp <span class="ot">&lt;-</span> <span class="fu">resize</span>(pop2, <span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">2000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="tidyverse-style-pipe-interface" class="level2">
<h2 class="anchored" data-anchor-id="tidyverse-style-pipe-interface">Tidyverse-style <a href="https://magrittr.tidyverse.org">pipe</a> <code>%&gt;%</code> interface</h2>
<p>A more concise way to express the same thing as before.</p>
<p><br></p>
<p>Step size decrease:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">how =</span> <span class="st">"step"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<p>Exponential increase:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">2000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="a-more-complex-model" class="level2">
<h2 class="anchored" data-anchor-id="a-more-complex-model">A more complex model</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">300</span>, <span class="at">parent =</span> pop1) <span class="sc">%&gt;%</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">1000</span>)</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>pop3 <span class="ot">&lt;-</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop3"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">400</span>, <span class="at">parent =</span> pop2) <span class="sc">%&gt;%</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">2500</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">800</span>)</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>pop4 <span class="ot">&lt;-</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop4"</span>, <span class="at">N =</span> <span class="dv">1500</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">parent =</span> pop3) <span class="sc">%&gt;%</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">700</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1200</span>, <span class="at">end =</span> <span class="dv">2000</span>)</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>pop5 <span class="ot">&lt;-</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop5"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">600</span>, <span class="at">parent =</span> pop4) <span class="sc">%&gt;%</span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">50</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">900</span>) <span class="sc">%&gt;%</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1600</span>, <span class="at">end =</span> <span class="dv">2200</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="each-object-carries-its-history" class="level2">
<h2 class="anchored" data-anchor-id="each-object-carries-its-history">Each object carries its history!</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>pop5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop5 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 600: split from pop4 (N = 100)
  - time 900: resize from 100 to 50 individuals
  - time 1600-2200: exponential resize from 50 to 1000 individuals</code></pre>
</div>
</div>
</section>
<section id="gene-flow-admixture" class="level2">
<h2 class="anchored" data-anchor-id="gene-flow-admixture">Gene flow / admixture</h2>
<p>We can schedule <code>gene_flow()</code> from <code>pop1</code> into <code>pop2</code> with:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">gene_flow</span>(<span class="at">from =</span> pop1, <span class="at">to =</span> pop2, <span class="at">start =</span> <span class="dv">2000</span>, <span class="at">end =</span> <span class="dv">2200</span>, <span class="at">rate =</span> <span class="fl">0.13</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>. . .</p>
<p>Here <code>rate = 0.13</code> means 13% migrants over the given time window will come from ‚Äúpop1‚Äù into ‚Äúpop2‚Äù.</p>
<p>. . .</p>
<p><br></p>
<p>Multiple gene-flow events can be gathered in a list:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> pop1, <span class="at">to =</span> pop2, <span class="at">start =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">600</span>, <span class="at">rate =</span> <span class="fl">0.13</span>),</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> ..., <span class="at">to =</span> ..., <span class="at">start =</span> ..., <span class="at">end =</span> ..., <span class="at">rate =</span> ...),</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  ...</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="last-step-before-simulation-compile_model" class="level2">
<h2 class="anchored" data-anchor-id="last-step-before-simulation-compile_model">Last step before simulation: <code>compile_model()</code></h2>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb20-2"><a href="#cb20-2"></a>  <span class="fu">list</span>(pop1, pop2, pop3, pop4, pop5),</span>
<span id="cb20-3"><a href="#cb20-3"></a>  <span class="at">generation_time =</span> <span class="dv">1</span>,</span>
<span id="cb20-4"><a href="#cb20-4"></a>  <span class="at">simulation_length =</span> <span class="dv">3000</span>,</span>
<span id="cb20-5"><a href="#cb20-5"></a>  <span class="at">direction =</span> <span class="st">"forward"</span></span>
<span id="cb20-6"><a href="#cb20-6"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
<strong>Compilation takes a list of model components, performs internal consistency checks, returns a single model object.</strong>
</center>
</section>
<section id="last-step-before-simulation-compile_model-1" class="level2">
<h2 class="anchored" data-anchor-id="last-step-before-simulation-compile_model-1">Last step before simulation: <code>compile_model()</code></h2>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb21-2"><a href="#cb21-2"></a>  <span class="fu">list</span>(pop1, pop2, pop3, pop4, pop5),</span>
<span id="cb21-3"><a href="#cb21-3"></a>  <span class="at">gene_flow =</span> gf,      <span class="co"># &lt;----- in case our model includes gene flow</span></span>
<span id="cb21-4"><a href="#cb21-4"></a>  <span class="at">generation_time =</span> <span class="dv">1</span>,</span>
<span id="cb21-5"><a href="#cb21-5"></a>  <span class="at">simulation_length =</span> <span class="dv">3000</span>,</span>
<span id="cb21-6"><a href="#cb21-6"></a>  <span class="at">direction =</span> <span class="st">"forward"</span></span>
<span id="cb21-7"><a href="#cb21-7"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
<strong>Gene flow(s) that we programmed are included</strong><br>via the <code>gene_flow</code> argument.
</center>
</section>
<section id="model-summary" class="level2">
<h2 class="anchored" data-anchor-id="model-summary">Model summary</h2>
<p>Typing the compiled <code>model</code> into R prints a brief summary:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'model' object 
--------------------- 
populations: pop1, pop2, pop3, pop4, pop5 
geneflow events: 1 
generation time: 1 
time direction: forward 
time units: generations 
total running length: 3000 time units
model type: non-spatial

configuration files in: /private/var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T/Rtmp1sJiEd/filefdfee4d924f_slendr_model </code></pre>
</div>
</div>
</section>
<section id="model-visualization" class="level2">
<h2 class="anchored" data-anchor-id="model-visualization">Model visualization</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model, <span class="at">proportions =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="slides_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="so-we-build-a-model" class="level1">
<h1>So we build a model</h1>
<center>
<h3 class="anchored">
How do we simulate data from it?
</h3>
</center>
<section id="built-in-simulation-engines" class="level2">
<h2 class="anchored" data-anchor-id="built-in-simulation-engines">Built-in ‚Äúsimulation engines‚Äù</h2>
<p><em>slendr</em> has two built-in simulation ‚Äúengine scripts‚Äù:</p>
<ul>
<li>SLiM engine (<a href="https://github.com/bodkan/slendr/blob/main/inst/scripts/script.slim">source</a>)</li>
<li><em>msprime</em> engine (<a href="https://github.com/bodkan/slendr/blob/main/inst/scripts/script.py">source</a>)</li>
</ul>
<p>They are designed to understand <em>slendr</em> models.</p>
<p>. . .</p>
<p><br>All you need to simulate data is this one line of code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<center>
<p><br></p>
<h3 class="anchored">
<strong>You don‚Äôt have to write <em>msprime</em> or SLiM code!</strong>
</h3>
</center>
</section>
</section>
<section id="the-result-of-a-slendr-simulation-is-a-tree-sequence-ts" class="level1">
<h1>The result of a <em>slendr</em> simulation is a <strong>tree sequence (<code>ts</code>)</strong></h1>
<section id="what-is-tree-sequence" class="level2">
<h2 class="anchored" data-anchor-id="what-is-tree-sequence">What is tree sequence?</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/tree_sequence_diagram_no_muts.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%"></p>
</figure>
</div>
<ul>
<li>a record of full genetic ancestry of a set of samples</li>
<li>an encoding of DNA sequence carried by those samples</li>
<li>an efficient analysis framework</li>
</ul>
</section>
</section>
<section id="why-tree-sequence" class="level1 page-columns page-full">
<h1>Why tree sequence?</h1>
<p><br></p>
<h3 class="anchored">
Why not VCF or a normal genotype table?
</h3>
<section id="what-we-usually-have" class="level2">
<h2 class="anchored" data-anchor-id="what-we-usually-have">What we usually have</h2>
<center>
<img src="images/vcf_screenshot.png" class="img-fluid" style="width:90.0%">
</center>
</section>
<section id="what-we-usually-want" class="level2">
<h2 class="anchored" data-anchor-id="what-we-usually-want">What we usually <em>want</em></h2>
<p>An understanding of our samples‚Äô evolutionary history:</p>
<center>
<img src="images/tree_sequence_diagram_no_muts.png" class="img-fluid">
</center>
<div class="fragment">
<center>
<h3 class="anchored">
<strong>This is exactly what a tree sequence <em>is</em>!</strong>
</h3>
</center>
</div>
</section>
<section id="the-magic-of-tree-sequences" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="the-magic-of-tree-sequences">The magic of tree sequences</h2>
<p>They allow computing of popgen statistics <em>without genotypes</em>!</p>
<center>
<img src="images/tree_sequence_diagram_no_muts.png" class="img-fluid">
</center>
<p>There is a ‚Äúduality‚Äù between mutations and branch lengths.</p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p>See an amazing paper by <a href="https://academic.oup.com/genetics/article/215/3/779/5930459">Ralph et al.&nbsp;(2020)</a> for more detail.</p>
</div></div></section>
<section id="what-if-we-need-mutations-though" class="level2">
<h2 class="anchored" data-anchor-id="what-if-we-need-mutations-though">What if we need mutations though?</h2>
<div class="fragment">
<p>Coalescent and mutation processes can be decoupled!</p>
<center>
<img src="images/tree_sequence_diagram_no_muts.png" class="img-fluid">
</center>
</div>
</section>
<section id="what-if-we-need-mutations-though-1" class="level2">
<h2 class="anchored" data-anchor-id="what-if-we-need-mutations-though-1">What if we need mutations though?</h2>
<p>Coalescent and mutation processes can be decoupled!</p>
<center>
<p><img src="images/tree_sequence_diagram.png" class="img-fluid"></p>
<div class="fragment">
<h3 class="anchored">
With <em>slendr</em>, we can add mutations <em>after</em> the simulation using <code>ts_mutate()</code>.
</h3>
</div>
</center>
<!-- ## Tree-sequence simulation in practice -->
<!-- Let' say we have a compiled `model` object. -->
<!-- Then, we can simulate a tree sequence like this: -->
<!-- ```{r} -->
<!-- ts <- msprime(model, sequence_length = 10e6, recombination_rate = 1e-8) -->
<!-- ``` -->
<!-- But we can also simulate mutations like this: -->
<!-- ```{r} -->
<!-- ts <- -->
<!--   msprime(model, sequence_length = 10e6, recombination_rate = 1e-8) %>% -->
<!--   ts_mutate(mutation_rate = 1e-8) -->
<!-- ``` -->
<!-- <br> -->
<!-- <center>We will be using `ts_mutate()` throughout.</center> -->
<!-- ## Tree sequences are _very_ efficient -->
<!-- <br> -->
<!-- This simulates 2 $\times$ 10000 chromosomes of 100 Mb: -->
<!-- ```{r} -->
<!-- #| eval: false -->
<!-- pop <- population("pop", time = 1e6, N = 10000) -->
<!-- model <- compile_model(pop, generation_time = 30, direction = "backward") -->
<!-- ts <- msprime(model, sequence_length = 100e6, recombination_rate = 1e-8) -->
<!-- ``` -->
<!-- . . . -->
<!-- <br> -->
<!-- **Runs in less than 30 seconds on my laptop!** -->
<!-- **Takes about 66 Mb of memory!** -->
<!-- ## How does this work?! -->
<!-- . . . -->
<!-- <center> -->
<!-- ![](images/tables.jpeg) -->
<!-- <center> -->
<!-- ## Tree-sequence tables -->
<!-- ::: row -->
<!-- ::: columns -->
<!-- ::: {.column width="60%"} -->
<!-- A tree (sequence) can be represented by -->
<!-- ::: incremental -->
<!-- -   a table of <font color="orange">n</font><font color="green">o</font><font color="darkblue">d</font><font color="green">e</font><font color="darkblue">s</font>, -->
<!-- -   a table of [edges]{.underline} between nodes, -->
<!-- -   a table of <font color="red">mutations</font> on edges -->
<!-- ::: -->
<!-- ::: -->
<!-- ::: {.column width="40%"} -->
<!-- <br> -->
<!-- <center>![](images/tree_diagram.png)</center> -->
<!-- ::: -->
<!-- ::: -->
<!-- ::: -->
<!-- . . . -->
<!-- <center> -->
<!-- <h3>**A set of such tables is a tree sequence.**</h3> -->
<!-- </center> -->
<!-- ::: aside -->
<!-- You can find much more information in  ([tskit docs](https://tskit.dev/tutorials/tables_and_editing.html)) -->
<!-- ::: -->
<!-- ## Tree-sequence tables in practice -->
<!-- ::: columns -->
<!-- ::: {.column width="50%"} -->
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- #| fig-height: 13 -->
<!-- set.seed(123) -->
<!-- ts <- msprime(model, sequence_length = 1e6, recombination_rate = 1e-8, random_seed = 42) %>% ts_mutate(1e-8, random_seed = 42) -->
<!-- # make a tiny example simplified tree sequence -->
<!-- ts_tiny <- ts_samples(ts) %>% sample_n(4) %>% pull(name) %>% ts_simplify(ts, simplify_to = .) -->
<!-- # extract tree #1 as an ape tree and also a tskit tree -->
<!-- t_phylo <- ts_phylo(ts_tiny, 1, quiet = TRUE) -->
<!-- t_tskit <- ts_tree(ts_tiny, 1) -->
<!-- # plot the phylo tree with ape -->
<!-- suppressPackageStartupMessages(library(ggtree)) -->
<!-- nodes <- ts_nodes(t_phylo) %>% as_tibble %>% dplyr::select(node = phylo_id, pop, node_id) -->
<!-- ggtree(t_phylo, branch.length = "none") %<+% nodes + -->
<!--   geom_label(aes(label = node_id), size = 15) +  -->
<!--   guides(color = "none") -->
<!-- ``` -->
<!-- ::: -->
<!-- ::: {.column width="50%"} -->
<!-- ::: fragment -->
<!-- nodes: -->
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- ts_nodes(t_phylo) %>% head(3) %>% .[, c("node_id", "pop_id", "time")] %>% as.data.frame() -->
<!-- ``` -->
<!-- ::: -->
<!-- ::: fragment -->
<!-- edges: -->
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- ts_edges(t_phylo) %>% head(3) %>% .[, c("child_node_id", "parent_node_id")] %>% as.data.frame() -->
<!-- ``` -->
<!-- ::: -->
<!-- ::: fragment -->
<!-- mutations: -->
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- ts_table(ts_tiny, "mutations") %>% filter(node %in% c(53, 22, 20, 74, 9)) %>% head(3) %>% .[, c("id", "site", "node", "time")] %>% as.data.frame() -->
<!-- ``` -->
<!-- ::: -->
<!-- ::: -->
<!-- ::: -->
</section>
<section id="lets-take-the-model-from-earlier" class="level2">
<h2 class="anchored" data-anchor-id="lets-take-the-model-from-earlier">Let‚Äôs take the <code>model</code> from earlier‚Ä¶</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="slides_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="and-simulate-data-from-it" class="level2">
<h2 class="anchored" data-anchor-id="and-simulate-data-from-it">‚Ä¶ and simulate data from it</h2>
<p><br></p>
<p>In our script we‚Äôll have something like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="sc">&lt;</span>... <span class="fu">population</span>() definitions ...<span class="sc">&gt;</span></span>
<span id="cb26-2"><a href="#cb26-2"></a></span>
<span id="cb26-3"><a href="#cb26-3"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(...)</span>
<span id="cb26-4"><a href="#cb26-4"></a>  </span>
<span id="cb26-5"><a href="#cb26-5"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb26-6"><a href="#cb26-6"></a>  <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">50e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="and-simulate-data-from-it-1" class="level2">
<h2 class="anchored" data-anchor-id="and-simulate-data-from-it-1">‚Ä¶ and simulate data from it</h2>
<p><br></p>
<p>In our script we‚Äôll have something like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="sc">&lt;</span>... <span class="fu">population</span>() definitions ...<span class="sc">&gt;</span></span>
<span id="cb27-2"><a href="#cb27-2"></a></span>
<span id="cb27-3"><a href="#cb27-3"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(...)</span>
<span id="cb27-4"><a href="#cb27-4"></a>  </span>
<span id="cb27-5"><a href="#cb27-5"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb27-6"><a href="#cb27-6"></a>  <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">50e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb27-7"><a href="#cb27-7"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p><br></p>
<center>
<h3 class="anchored">
Always use <code>ts_mutate()</code> during exercises! Otherwise you‚Äôll get weird results due to the lack of mutations on the tree sequence.
</h3>
</center>
</div>
</section>
</section>
<section id="so-we-can-simulate-data" class="level1">
<h1>So we can simulate data</h1>
<center>
<h3 class="anchored">
How do we work with this <code>ts</code> thing?
</h3>
</center>
<section id="standard-genotype-formats" class="level2">
<h2 class="anchored" data-anchor-id="standard-genotype-formats">Standard genotype formats</h2>
<p>If a tree sequence doesn‚Äôt cut it, you can always:</p>
<ul>
<li>export genotypes to a VCF file:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_vcf</span>(ts, <span class="at">path =</span> <span class="st">"path/to/a/file.vcf.gz"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>export genotypes in an EIGENSTRAT format:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_eigenstrat</span>(ts, <span class="at">prefix =</span> <span class="st">"path/to/eigenstrat/prefix"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>access genotypes in a data frame:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_genotypes</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>2 multiallelic sites (0.122% out of 1642 total) detected and removed</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   pos pop1_1_chr1 pop1_1_chr2 pop1_2_chr1 pop1_2_chr2 pop1_3_chr1 pop1_3_chr2
1  212           0           0           0           0           0           0
2 1506           0           0           0           0           0           0</code></pre>
</div>
</div>
</section>
<section id="slendrs-r-interface-to-tskit-statistics" class="level2">
<h2 class="anchored" data-anchor-id="slendrs-r-interface-to-tskit-statistics"><em>slendr</em>‚Äôs R interface to <a href="https://tskit.dev/tskit"><em>tskit</em></a> statistics</h2>
<center>
<img src="images/slendr_tskit.png" class="img-fluid">
</center>
<p>Allele-frequecy spectrum, diversity <span class="math inline">\(\pi\)</span>, <span class="math inline">\(F_{ST}\)</span>, Tajima‚Äôs D, etc.</p>
<p><strong>Find help at <a href="https://slendr.net/reference">slendr.net/reference</a> or in R under <code>?ts_fst</code> etc.</strong></p>
</section>
<section id="specifying-individuals-to-compute-on" class="level2">
<h2 class="anchored" data-anchor-id="specifying-individuals-to-compute-on">Specifying individuals to compute on</h2>
<p>We can get samples recorded in <code>ts</code> with <code>ts_samples()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 √ó 3
  name    time pop  
  &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
1 pop1_1  3001 pop1 
2 pop1_2  3001 pop1 </code></pre>
</div>
</div>
<p>. . .</p>
<p>A shortcut <code>ts_names()</code> can also be useful:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_names</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pop1_1" "pop1_2" "pop1_3" "pop1_4" "pop1_5"</code></pre>
</div>
</div>
<p>. . .</p>
<p>We can also get a per-population list of individuals:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>$pop1
[1] "pop1_824" "pop1_11"  "pop1_873" "pop1_317" "pop1_193"</code></pre>
</div>
</div>
</section>
</section>
<section id="section-5" class="level1">
<h1></h1>
<center>
<h2 class="anchored" data-anchor-id="section-5">
All <em>slendr</em> statistics accept this as their second (and further) arguments.
</h2>
</center>
<p><br></p>
<center>
This is modelled after the <code>sample_sets=</code> argument of the respective <a href="https://tskit.dev/tskit/docs/stable/python-api.html#statistics"><em>tskit</em> Python methods</a> (except that you use ‚Äúsymbolic names‚Äù of individuals, not tree-sequence node numbers).
</center>
<section id="example-nucleotide-diversity" class="level2">
<h2 class="anchored" data-anchor-id="example-nucleotide-diversity">Example: <a href="https://en.wikipedia.org/wiki/Nucleotide_diversity">nucleotide diversity</a></h2>
<div class="columns">
<div class="column" style="width:45%;">
<p>Get list of individuals in each population:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pop1" "pop2"</code></pre>
</div>
</div>
<p><br></p>
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>samples<span class="sc">$</span>pop1 <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pop1_1" "pop1_2" "pop1_3"</code></pre>
</div>
</div>
</div>
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>samples<span class="sc">$</span>pop2 <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pop2_1" "pop2_2" "pop2_3"</code></pre>
</div>
</div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:53%;">
<div class="fragment">
<p>Compute nucleotide diversity:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_diversity</span>(ts, <span class="at">sample_sets =</span> samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 √ó 2
  set   diversity
  &lt;chr&gt;     &lt;dbl&gt;
1 pop1  0.000400 
2 pop2  0.0000713</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="example-allele-frequency-spectrum" class="level2">
<h2 class="anchored" data-anchor-id="example-allele-frequency-spectrum">Example: <a href="https://en.wikipedia.org/wiki/Allele_frequency_spectrum">allele frequency spectrum</a></h2>
<div class="columns">
<div class="column" style="width:45%;">
<p>Get names of all individuals:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pop_1" "pop_2" "pop_3" "pop_4" "pop_5"</code></pre>
</div>
</div>
<div class="fragment">
<p><br></p>
<p>Compute the AFS:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>afs <span class="ot">&lt;-</span> <span class="fu">ts_afs</span>(ts, <span class="at">sample_sets =</span> <span class="fu">list</span>(samples))</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>afs[<span class="sc">-</span><span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 3917 2151 1432  941  740  624  607  587  416  385</code></pre>
</div>
</div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:53%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(afs[<span class="sc">-</span><span class="dv">1</span>], <span class="at">type =</span> <span class="st">"b"</span>,</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"allele count bin"</span>,</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slides_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="working-with-multi-population-statistics" class="level2">
<h2 class="anchored" data-anchor-id="working-with-multi-population-statistics">Working with ‚Äúmulti-population‚Äù statistics</h2>
<p>For multipopulation statistics, you need a (named) list of samples recorded in each population.</p>
<p><code>ts_names()</code> has a <code>split = "pop"</code> option just for this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>$&lt;NA&gt;
NULL

$&lt;NA&gt;
NULL</code></pre>
</div>
</div>
<p>You can use this in place of <code>sample_names</code> in code like:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_diversity</span>(ts, <span class="at">sample_sets =</span> samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="more-information" class="level2">
<h2 class="anchored" data-anchor-id="more-information">More information</h2>
<p><br></p>
<ul>
<li><p><em>slendr</em> paper is now in <a href="https://evolbiol.peercommunityin.org/articles/rec?id=630">PCI EvolBiol</a></p></li>
<li><p>documentation, tutorials is <a href="https://www.slendr.net">here</a></p></li>
<li><p>GitHub repo (bug reports!) is <a href="http://github.com/bodkan/slendr">here</a></p></li>
</ul>
<hr>
<ul>
<li>check out my new <a href="http://github.com/bodkan/demografr"><em>demografr</em></a> inference package</li>
</ul>
<hr>
<ul>
<li>contact details at <a href="https://bodkan.net">bodkan.net</a></li>
</ul>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "Óßã";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>